<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Normalization</title>
    <meta charset="utf-8" />
    <meta name="author" content="Leonardo Collado-Torres" />
    <meta name="date" content="2020-03-23" />
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-161558379-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
    
      gtag('config', 'UA-161558379-1');
    </script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <strong>Normalization</strong>
## Analyzing <strong>scRNA-seq</strong> data with <strong>Bioconductor</strong> for <strong>LCG-EJ-UNAM</strong> March 2020
### <a href="http://lcolladotor.github.io/">Leonardo Collado-Torres</a>
### 2020-03-23

---






class: inverse

.center[

&lt;a href="https://bioconductor.org/"&gt;&lt;img src="https://osca.bioconductor.org/cover.png" style="width: 30%"/&gt;&lt;/a&gt;

&lt;a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"&gt;&lt;img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /&gt;&lt;/a&gt;&lt;br /&gt;This work is licensed under a &lt;a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"&gt;Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License&lt;/a&gt;.

&lt;a href='https://clustrmaps.com/site/1b5pl'  title='Visit tracker'&gt;&lt;img src='//clustrmaps.com/map_v2.png?cl=ffffff&amp;w=150&amp;t=n&amp;d=tq5q8216epOrQBSllNIKhXOHUHi-i38brzUURkQEiXw'/&gt;&lt;/a&gt;

]

.footnote[ Download the materials for this course with `usethis::use_course('lcolladotor/osca_LIIGH_UNAM_2020')` or view online at [**lcolladotor.github.io/osca_LIIGH_UNAM_2020**](http://lcolladotor.github.io/osca_LIIGH_UNAM_2020).]

&lt;style type="text/css"&gt;
/* From https://github.com/yihui/xaringan/issues/147  */
.scroll-output {
  height: 80%;
  overflow-y: scroll;
}

/* https://stackoverflow.com/questions/50919104/horizontally-scrollable-output-on-xaringan-slides */
pre {
  max-width: 100%;
  overflow-x: scroll;
}

/* From https://github.com/yihui/xaringan/wiki/Font-Size */
.tiny{
  font-size: 40%
}

/* From https://github.com/yihui/xaringan/wiki/Title-slide */
.title-slide {
  background-image: url(https://raw.githubusercontent.com/Bioconductor/OrchestratingSingleCellAnalysis/master/images/Workflow.png);
  background-size: 33%;
  background-position: 0% 100%
}
&lt;/style&gt;

---

# Slides by Peter Hickey

View them [here](https://docs.google.com/presentation/d/1_tCNLiEsQ_TgsqHHf9_1lzXSaM_LunEHxBq3k130dQI/edit#slide=id.g7cc450648d_0_118)

---

# Code and output

.scroll-output[


```r
library('scRNAseq')
sce.zeisel &lt;- ZeiselBrainData(ensembl = TRUE)
```

```
## snapshotDate(): 2019-10-22
```

```
## see ?scRNAseq and browseVignettes('scRNAseq') for documentation
```

```
## loading from cache
```

```
## see ?scRNAseq and browseVignettes('scRNAseq') for documentation
```

```
## loading from cache
```

```
## see ?scRNAseq and browseVignettes('scRNAseq') for documentation
```

```
## loading from cache
```

```
## snapshotDate(): 2019-10-29
```

```
## loading from cache
```

```
## Warning: Unable to map 1565 of 20006 requested IDs.
```

```r
# Quality control
library('scater')
is.mito &lt;- which(rowData(sce.zeisel)$featureType == "mito")
stats &lt;- perCellQCMetrics(sce.zeisel, subsets = list(Mt = is.mito))
qc &lt;-
    quickPerCellQC(stats,
        percent_subsets = c("altexps_ERCC_percent", "subsets_Mt_percent"))
sce.zeisel &lt;- sce.zeisel[, !qc$discard]
```

]

---

.scroll-output[


```r
# Library size factors
lib.sf.zeisel &lt;- librarySizeFactors(sce.zeisel)

# Examine distribution of size factors
summary(lib.sf.zeisel)
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.1754  0.5682  0.8669  1.0000  1.2758  4.0651
```

```r
hist(log10(lib.sf.zeisel), xlab = "Log10[Size factor]", col = "grey80")
```

![](04-normalization_files/figure-html/all_code2-1.png)&lt;!-- --&gt;

```r
ls.zeisel &lt;- colSums(counts(sce.zeisel))
plot(
    ls.zeisel,
    lib.sf.zeisel,
    log = "xy",
    xlab = "Library size",
    ylab = "Size factor"
)
```

![](04-normalization_files/figure-html/all_code2-2.png)&lt;!-- --&gt;

]


---

# Exercise


--

* Are `ls.zeisel` and `lib.sf.zeisel` identical?

--

* Are they proportional?

--

* Compute `lib.sf.zeisel` manually


---

# Solution

--

* Check the **Details** at `?scater::librarySizeFactors`

--

* Compute the size factors manually


```r
## First compute the sums
zeisel_sums &lt;- colSums(counts(sce.zeisel))
identical(zeisel_sums, ls.zeisel)
```

```
## [1] TRUE
```

```r
## Next, make them have unity mean
zeisel_size_factors &lt;- zeisel_sums/mean(zeisel_sums)
identical(zeisel_size_factors, lib.sf.zeisel)
```

```
## [1] TRUE
```

--

* Check the [source code](https://github.com/davismcc/scater/blob/master/R/librarySizeFactors.R)

---

.scroll-output[


```r
# Normalization by convolution

library('scran')
# Pre-clustering
set.seed(100)
clust.zeisel &lt;- quickCluster(sce.zeisel)
# Compute deconvolution size factors
deconv.sf.zeisel &lt;-
    calculateSumFactors(sce.zeisel, clusters = clust.zeisel, min.mean = 0.1)

# Examine distribution of size factors
summary(deconv.sf.zeisel)
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.1282  0.4859  0.8248  1.0000  1.3194  4.6521
```

```r
hist(log10(deconv.sf.zeisel), xlab = "Log10[Size factor]",
    col = "grey80")
```

![](04-normalization_files/figure-html/all_code3-1.png)&lt;!-- --&gt;

```r
plot(
    ls.zeisel,
    deconv.sf.zeisel,
    log = "xy",
    xlab = "Library size",
    ylab = "Size factor"
)
```

![](04-normalization_files/figure-html/all_code3-2.png)&lt;!-- --&gt;

]

---

# Exercises

--

* How many quick clusters did we get? 

--

* How many cells per quick cluster did we get?

--

* How many quick clusters will we get if we set the minimum size to 200? Use 100 as the seed.

--

* How many lines do you see?


???

* 12
* From 113 to 325, `sort(table(clust.zeisel))`
* 10 `set.seed(100); sort(table(quickCluster(sce.zeisel, min.size = 200)))` 
* Several ones near the diagonal. They are potentially 7 `table(factor(sce.zeisel$level1class))`

---

.scroll-output[


```r
# Library size factors vs. convolution size factors

# Colouring points using the supplied cell-types
plot(
    lib.sf.zeisel,
    deconv.sf.zeisel,
    xlab = "Library size factor",
    ylab = "Deconvolution size factor",
    log = 'xy',
    pch = 16,
    col = as.integer(factor(sce.zeisel$level1class))
)
abline(a = 0, b = 1, col = "red")
```

![](04-normalization_files/figure-html/all_code4-1.png)&lt;!-- --&gt;

]

---

class: middle

.center[

# Thanks!

Slides created via the R package [**xaringan**](https://github.com/yihui/xaringan) and themed with [**xaringanthemer**](https://github.com/gadenbuie/xaringanthemer).

This course is based on the book [**Orchestrating Single Cell Analysis with Bioconductor**](https://osca.bioconductor.org/) by [Aaron Lun](https://www.linkedin.com/in/aaron-lun-869b5894/), [Robert Amezquita](https://robertamezquita.github.io/), [Stephanie Hicks](https://www.stephaniehicks.com/) and [Raphael Gottardo](http://rglab.org), plus [**WEHI's scRNA-seq course**](https://drive.google.com/drive/folders/1cn5d-Ey7-kkMiex8-74qxvxtCQT6o72h) by [Peter Hickey](https://www.peterhickey.org/).


You can find the files for this course at [lcolladotor/osca_LIIGH_UNAM_2020](https://github.com/lcolladotor/osca_LIIGH_UNAM_2020). 

Instructor: [**Leonardo Collado-Torres**](http://lcolladotor.github.io/).

&lt;a href="https://www.libd.org"&gt;&lt;img src="img/LIBD_logo.jpg" style="width: 20%" /&gt;&lt;/a&gt;

]

.footnote[ Download the materials for this course with `usethis::use_course('lcolladotor/osca_LIIGH_UNAM_2020')` or view online at [**lcolladotor.github.io/osca_LIIGH_UNAM_2020**](http://lcolladotor.github.io/osca_LIIGH_UNAM_2020).]

---

# R session information


.scroll-output[
.tiny[


```r
options(width = 120)
sessioninfo::session_info()
```

```
## ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.6.3 (2020-02-29)
##  os       macOS Catalina 10.15.3      
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  ctype    en_US.UTF-8                 
##  tz       America/New_York            
##  date     2020-03-23                  
## 
## ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
##  package                * version  date       lib source                                   
##  AnnotationDbi          * 1.48.0   2019-10-29 [1] Bioconductor                             
##  AnnotationFilter       * 1.10.0   2019-10-29 [1] Bioconductor                             
##  AnnotationHub            2.18.0   2019-10-29 [1] Bioconductor                             
##  askpass                  1.1      2019-01-13 [1] CRAN (R 3.6.0)                           
##  assertthat               0.2.1    2019-03-21 [1] CRAN (R 3.6.0)                           
##  beeswarm                 0.2.3    2016-04-25 [1] CRAN (R 3.6.0)                           
##  Biobase                * 2.46.0   2019-10-29 [1] Bioconductor                             
##  BiocFileCache            1.10.2   2019-11-08 [1] Bioconductor                             
##  BiocGenerics           * 0.32.0   2019-10-29 [1] Bioconductor                             
##  BiocManager              1.30.10  2019-11-16 [1] CRAN (R 3.6.1)                           
##  BiocNeighbors            1.4.2    2020-02-29 [1] Bioconductor                             
##  BiocParallel           * 1.20.1   2019-12-21 [1] Bioconductor                             
##  BiocSingular             1.2.2    2020-02-14 [1] Bioconductor                             
##  BiocVersion              3.10.1   2019-06-06 [1] Bioconductor                             
##  biomaRt                  2.42.0   2019-10-29 [1] Bioconductor                             
##  Biostrings               2.54.0   2019-10-29 [1] Bioconductor                             
##  bit                      1.1-15.2 2020-02-10 [1] CRAN (R 3.6.0)                           
##  bit64                    0.9-7    2017-05-08 [1] CRAN (R 3.6.0)                           
##  bitops                   1.0-6    2013-08-17 [1] CRAN (R 3.6.0)                           
##  blob                     1.2.1    2020-01-20 [1] CRAN (R 3.6.0)                           
##  cli                      2.0.2    2020-02-28 [1] CRAN (R 3.6.0)                           
##  codetools                0.2-16   2018-12-24 [1] CRAN (R 3.6.3)                           
##  colorout               * 1.2-1    2019-05-07 [1] Github (jalvesaq/colorout@7ea9440)       
##  colorspace               1.4-1    2019-03-18 [1] CRAN (R 3.6.0)                           
##  crayon                   1.3.4    2017-09-16 [1] CRAN (R 3.6.0)                           
##  curl                     4.3      2019-12-02 [1] CRAN (R 3.6.0)                           
##  DBI                      1.1.0    2019-12-15 [1] CRAN (R 3.6.0)                           
##  dbplyr                   1.4.2    2019-06-17 [1] CRAN (R 3.6.0)                           
##  DelayedArray           * 0.12.2   2020-01-06 [1] Bioconductor                             
##  DelayedMatrixStats       1.8.0    2019-10-29 [1] Bioconductor                             
##  digest                   0.6.25   2020-02-23 [1] CRAN (R 3.6.0)                           
##  dplyr                    0.8.5    2020-03-07 [1] CRAN (R 3.6.0)                           
##  dqrng                    0.2.1    2019-05-17 [1] CRAN (R 3.6.0)                           
##  edgeR                    3.28.1   2020-02-26 [1] Bioconductor                             
##  ensembldb              * 2.10.2   2019-11-20 [1] Bioconductor                             
##  evaluate                 0.14     2019-05-28 [1] CRAN (R 3.6.0)                           
##  ExperimentHub            1.12.0   2019-10-29 [1] Bioconductor                             
##  fansi                    0.4.1    2020-01-08 [1] CRAN (R 3.6.0)                           
##  fastmap                  1.0.1    2019-10-08 [1] CRAN (R 3.6.0)                           
##  GenomeInfoDb           * 1.22.0   2019-10-29 [1] Bioconductor                             
##  GenomeInfoDbData         1.2.2    2019-10-31 [1] Bioconductor                             
##  GenomicAlignments        1.22.1   2019-11-12 [1] Bioconductor                             
##  GenomicFeatures        * 1.38.2   2020-02-15 [1] Bioconductor                             
##  GenomicRanges          * 1.38.0   2019-10-29 [1] Bioconductor                             
##  ggbeeswarm               0.6.0    2017-08-07 [1] CRAN (R 3.6.0)                           
##  ggplot2                * 3.3.0    2020-03-05 [1] CRAN (R 3.6.0)                           
##  glue                     1.3.2    2020-03-12 [1] CRAN (R 3.6.0)                           
##  gridExtra                2.3      2017-09-09 [1] CRAN (R 3.6.0)                           
##  gtable                   0.3.0    2019-03-25 [1] CRAN (R 3.6.0)                           
##  hms                      0.5.3    2020-01-08 [1] CRAN (R 3.6.0)                           
##  htmltools                0.4.0    2019-10-04 [1] CRAN (R 3.6.0)                           
##  httpuv                   1.5.2    2019-09-11 [1] CRAN (R 3.6.0)                           
##  httr                     1.4.1    2019-08-05 [1] CRAN (R 3.6.0)                           
##  igraph                   1.2.5    2020-03-19 [1] CRAN (R 3.6.0)                           
##  interactiveDisplayBase   1.24.0   2019-10-29 [1] Bioconductor                             
##  IRanges                * 2.20.2   2020-01-13 [1] Bioconductor                             
##  irlba                    2.3.3    2019-02-05 [1] CRAN (R 3.6.0)                           
##  knitr                    1.28     2020-02-06 [1] CRAN (R 3.6.0)                           
##  later                    1.0.0    2019-10-04 [1] CRAN (R 3.6.0)                           
##  lattice                  0.20-40  2020-02-19 [1] CRAN (R 3.6.0)                           
##  lazyeval                 0.2.2    2019-03-15 [1] CRAN (R 3.6.0)                           
##  lifecycle                0.2.0    2020-03-06 [1] CRAN (R 3.6.0)                           
##  limma                    3.42.2   2020-02-03 [1] Bioconductor                             
##  locfit                   1.5-9.1  2013-04-20 [1] CRAN (R 3.6.0)                           
##  magrittr                 1.5      2014-11-22 [1] CRAN (R 3.6.0)                           
##  Matrix                   1.2-18   2019-11-27 [1] CRAN (R 3.6.3)                           
##  matrixStats            * 0.56.0   2020-03-13 [1] CRAN (R 3.6.0)                           
##  memoise                  1.1.0    2017-04-21 [1] CRAN (R 3.6.0)                           
##  mime                     0.9      2020-02-04 [1] CRAN (R 3.6.0)                           
##  munsell                  0.5.0    2018-06-12 [1] CRAN (R 3.6.0)                           
##  openssl                  1.4.1    2019-07-18 [1] CRAN (R 3.6.0)                           
##  pillar                   1.4.3    2019-12-20 [1] CRAN (R 3.6.0)                           
##  pkgconfig                2.0.3    2019-09-22 [1] CRAN (R 3.6.1)                           
##  prettyunits              1.1.1    2020-01-24 [1] CRAN (R 3.6.2)                           
##  progress                 1.2.2    2019-05-16 [1] CRAN (R 3.6.0)                           
##  promises                 1.1.0    2019-10-04 [1] CRAN (R 3.6.0)                           
##  ProtGenerics             1.18.0   2019-10-29 [1] Bioconductor                             
##  purrr                    0.3.3    2019-10-18 [1] CRAN (R 3.6.0)                           
##  R6                       2.4.1    2019-11-12 [1] CRAN (R 3.6.1)                           
##  rappdirs                 0.3.1    2016-03-28 [1] CRAN (R 3.6.0)                           
##  Rcpp                     1.0.4    2020-03-17 [1] CRAN (R 3.6.0)                           
##  RCurl                    1.98-1.1 2020-01-19 [1] CRAN (R 3.6.0)                           
##  rlang                    0.4.5    2020-03-01 [1] CRAN (R 3.6.0)                           
##  rmarkdown                2.1      2020-01-20 [1] CRAN (R 3.6.0)                           
##  Rsamtools                2.2.3    2020-02-23 [1] Bioconductor                             
##  RSQLite                  2.2.0    2020-01-07 [1] CRAN (R 3.6.0)                           
##  rstudioapi               0.11     2020-02-07 [1] CRAN (R 3.6.0)                           
##  rsvd                     1.0.3    2020-02-17 [1] CRAN (R 3.6.0)                           
##  rtracklayer              1.46.0   2019-10-29 [1] Bioconductor                             
##  S4Vectors              * 0.24.3   2020-01-18 [1] Bioconductor                             
##  scales                   1.1.0    2019-11-18 [1] CRAN (R 3.6.1)                           
##  scater                 * 1.14.6   2019-12-16 [1] Bioconductor                             
##  scran                  * 1.14.6   2020-02-03 [1] Bioconductor                             
##  scRNAseq               * 2.0.2    2019-11-12 [1] Bioconductor                             
##  sessioninfo              1.1.1    2018-11-05 [1] CRAN (R 3.6.0)                           
##  shiny                    1.4.0.2  2020-03-13 [1] CRAN (R 3.6.0)                           
##  SingleCellExperiment   * 1.8.0    2019-10-29 [1] Bioconductor                             
##  statmod                  1.4.34   2020-02-17 [1] CRAN (R 3.6.0)                           
##  stringi                  1.4.6    2020-02-17 [1] CRAN (R 3.6.0)                           
##  stringr                  1.4.0    2019-02-10 [1] CRAN (R 3.6.0)                           
##  SummarizedExperiment   * 1.16.1   2019-12-19 [1] Bioconductor                             
##  tibble                   2.1.3    2019-06-06 [1] CRAN (R 3.6.0)                           
##  tidyselect               1.0.0    2020-01-27 [1] CRAN (R 3.6.0)                           
##  vctrs                    0.2.4    2020-03-10 [1] CRAN (R 3.6.0)                           
##  vipor                    0.4.5    2017-03-22 [1] CRAN (R 3.6.0)                           
##  viridis                  0.5.1    2018-03-29 [1] CRAN (R 3.6.0)                           
##  viridisLite              0.3.0    2018-02-01 [1] CRAN (R 3.6.0)                           
##  withr                    2.1.2    2018-03-15 [1] CRAN (R 3.6.0)                           
##  xaringan                 0.15     2020-03-04 [1] CRAN (R 3.6.3)                           
##  xaringanthemer         * 0.2.0    2020-03-22 [1] Github (gadenbuie/xaringanthemer@460f441)
##  xfun                     0.12     2020-01-13 [1] CRAN (R 3.6.0)                           
##  XML                      3.99-0.3 2020-01-20 [1] CRAN (R 3.6.0)                           
##  xtable                   1.8-4    2019-04-21 [1] CRAN (R 3.6.0)                           
##  XVector                  0.26.0   2019-10-29 [1] Bioconductor                             
##  yaml                     2.2.1    2020-02-01 [1] CRAN (R 3.6.0)                           
##  zlibbioc                 1.32.0   2019-10-29 [1] Bioconductor                             
## 
## [1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```


]]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  /* Replace <script> tags in slides area to make them executable
   *
   * Runs after post-processing of markdown source into slides and replaces only
   * <script>s on the last slide of continued slides using the .has-continuation
   * class added by xaringan. Finally, any <script>s in the slides area that
   * aren't executed are commented out.
   */
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container:not(.has-continuation) script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
  var scriptsNotExecuted = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container.has-continuation script'
  );
  if (!scriptsNotExecuted.length) return;
  for (var i = 0; i < scriptsNotExecuted.length; i++) {
    var comment = document.createComment(scriptsNotExecuted[i].outerHTML)
    scriptsNotExecuted[i].parentElement.replaceChild(comment, scriptsNotExecuted[i])
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
